sensing/pointcloud_preprocessor/src/blockage_diag/blockage_diag_nodelet.cpp

```diff
@@ -0,0 +1,243 @@
+  stat.add("sky_blockage_ratio", std::to_string(sky_blockage_ratio_));
+  stat.add("sky_blockage_count", std::to_string(sky_blockage_count_));
+  stat.add(
+    "sky_blockage_range_deg", "[" + std::to_string(sky_blockage_range_deg_[0]) + "," +
+                                std::to_string(sky_blockage_range_deg_[1]) + "]");
+
+  // TODO(badai-nguyen): consider sky_blockage_ratio_ for DiagnosticsStatus." [todo]
+
+  auto level = DiagnosticStatus::OK;
+  if (ground_blockage_ratio_ < 0) {
+    level = DiagnosticStatus::STALE;
+  } else if (
+    (ground_blockage_ratio_ > blockage_ratio_threshold_) &&
+    (ground_blockage_count_ > blockage_count_threshold_)) {
+    level = DiagnosticStatus::ERROR;
+  } else if (ground_blockage_ratio_ > 0.0f) {
+    level = DiagnosticStatus::WARN;
+  } else {
+    level = DiagnosticStatus::OK;
+  }
+
+  std::string msg;
+  if (level == DiagnosticStatus::OK) {
+    msg = "OK";
+  } else if (level == DiagnosticStatus::WARN) {
+    msg = "WARNING: LiDAR ground blockage";
+  } else if (level == DiagnosticStatus::ERROR) {
+    msg = "ERROR: LiDAR ground blockage";
+  } else if (level == DiagnosticStatus::STALE) {
+    msg = "STALE";
+  }
+  stat.summary(level, msg);
+}
+
+void BlockageDiagComponent::filter(
+  const PointCloud2ConstPtr & input, [[maybe_unused]] const IndicesPtr & indices,
+  PointCloud2 & output)
+{
+  boost::mutex::scoped_lock lock(mutex_);
+  uint horizontal_bins = static_cast<uint>((angle_range_deg_[1] - angle_range_deg_[0]));
+  uint vertical_bins = vertical_bins_;
+  pcl::PointCloud<PointXYZIRADRT>::Ptr pcl_input(new pcl::PointCloud<PointXYZIRADRT>);
+  pcl::fromROSMsg(*input, *pcl_input);
+  cv::Mat lidar_depth_map(cv::Size(horizontal_bins, vertical_bins), CV_16UC1, cv::Scalar(0));
+  cv::Mat lidar_depth_map_8u(cv::Size(horizontal_bins, vertical_bins), CV_8UC1, cv::Scalar(0));
+  if (pcl_input->points.empty()) {
+    ground_blockage_ratio_ = 1.0f;
+    sky_blockage_ratio_ = 1.0f;
+    ground_blockage_count_ += 1;
+    sky_blockage_count_ += 1;
+    ground_blockage_range_deg_[0] = angle_range_deg_[0];
+    ground_blockage_range_deg_[1] = angle_range_deg_[1];
+    sky_blockage_range_deg_[0] = angle_range_deg_[0];
+    sky_blockage_range_deg_[1] = angle_range_deg_[1];
+  } else {
+    for (const auto & p : pcl_input->points) {
+      if ((p.azimuth / 100.0 > angle_range_deg_[0]) && (p.azimuth / 100.0 < angle_range_deg_[1])) {
+        switch (lidar_model_map_[lidar_model_]) {
+          case 0:  // Channel id increases from top to bottom: Pandar40P
+            lidar_depth_map.at<uint16_t>(
+              p.ring, static_cast<uint>((p.azimuth / 100.0 - angle_range_deg_[0]))) +=
+              static_cast<uint16_t>(6250.0 / p.distance);  // make image clearly
+            lidar_depth_map_8u.at<uint8_t>(
+              p.ring, static_cast<uint>((p.azimuth / 100.0 - angle_range_deg_[0]))) = 255;
+            break;
+          case 1:  // Channel id decreases from top to bottom: PandarQT
```

Comment:
[IMO] using `else if (lidar_model_ == "PandarQT")` is more understandable code rather than `case 1`